_posts/2024/2024-02-06-about-caching-in-eo.md

```diff
@@ -0,0 +1,263 @@
+---
+layout: post
+date: 2024-02-06
+title: "Build cache in EO and other build systems"
+author: Alekseeva Yana
+---
+
+
+## Introduction 
+In [EO](https://github.com/objectionary/eo), caching is used to speed up program compilation.
+Recently we found a caching 
+[bug](https://github.com/objectionary/eo/issues/2790) in `eo-maven-plugin`
+for EO version `0.34.0`. The bug occurred because the old verification method
+contains a comparison of the compilation time and caching time to search for the cached file.
+This is not the most reliable verification method,
+because caching time does not have to be equal to compilation time.
+We came to the conclusion that we need caching with a reliable verification method.
+And this verification method should not use the information that the cached file contains.
+
+The goal of this blog is to research caching in frequently used build systems (`ccache`, `Maven`, `Gradle`)
+and to implement effective caching in [EO](https://github.com/objectionary/eo).
+
+<!--more-->
+
+## Build caching of existing build systems
+
+### ccache/sccache
+In compiled programming languages, building a project containing many source code files takes a long time.
+This time is spent on loading of libraries, preparing, optimizing, checking the code, and so on.
+To speed up the assembly of compiled languages, [ccache](https://ccache.dev) 
+and [sccache](https://github.com/mozilla/sccache) are used.
+Let's look at the assembly scheme using C++ as an example
+to imagine the build process in compiled languages:
+
+<p align="center">
+  <img src="/images/defaultCPhase.svg">
+</p>
+
+1) First, preprocessor gets the input files. The input files are source files `.cpp` and header files `.h`.
+The result is a single edited file `.cpp` with human-readable code that the compiler will get.
+
+
+2) The compiler receives the edited code file `.cpp` and converts it into machine code, presented in an object file.
+At the compilation stage, parsing occurs, which checks whether the code matches
+rules of a specific programming language.
+At the end, the compiler optimizes the resulting machine code and produces an object file. 
+To speed up compilation, different files of the same project are compiled in parallel.
+
+3)  Then, the [Linker](https://en.wikipedia.org/wiki/Linker_(computing)) gets object files.
+The result of the linker is an executable `.exe` file.
+
+
+As a result, in compiled languages, multiple files are simultaneously and independently converted
+into machine code at the compilation stage.
+This machine code is then combined into one executable file.
+
+
+`ccache` hash algorithm, for the hashing of information to find cached files fast. 
```

Comment:
@Yanich96 This sentence is inconsistent with the previous one. Moreover, it seems we have grammar errors here.

Reply:
@volodya-lombrozo the two last sentences (lines 52-55) was unnecessary. I need delete them. Thanks, I have found grammar error.